cdap-app-fabric/src/main/java/co/cask/cdap/app/guice/ImpersonatedTwillRunnerService.java

```diff
@@ -0,0 +1,183 @@
+  }
+
+  @Override
+  public TwillController lookup(String applicationName, RunId runId) {
+    TwillController controller = delegate.lookup(applicationName, runId);
+    if (isMasterService(applicationName)) {
+      return controller;
+    }
+    try {
+      return new ImpersonatedTwillController(controller, impersonator, TwillAppNames.fromTwillAppName(applicationName));
+    } catch (IllegalArgumentException e) {
+      // If the conversion from twill app name to programId failed, don't wrap
+      return controller;
+    }
+  }
+
+  @Override
+  public Iterable<TwillController> lookup(String applicationName) {
+    return wrapControllers(delegate.lookup(applicationName), applicationName);
+  }
+
+  @Override
+  public Iterable<LiveInfo> lookupLive() {
+    return Iterables.transform(delegate.lookupLive(), new Function<LiveInfo, LiveInfo>() {
+      @Override
+      public LiveInfo apply(final LiveInfo liveInfo) {
+        return new LiveInfo() {
+          @Override
+          public String getApplicationName() {
+            return liveInfo.getApplicationName();
+          }
+
+          @Override
+          public Iterable<TwillController> getControllers() {
+            return wrapControllers(liveInfo.getControllers(), liveInfo.getApplicationName());
+          }
+        };
+      }
+    });
+  }
+
+  @Override
+  public Cancellable scheduleSecureStoreUpdate(SecureStoreUpdater updater, long initialDelay,
+                                               long delay, TimeUnit unit) {
+    return delegate.scheduleSecureStoreUpdate(wrapSecureStoreUpdater(updater), initialDelay, delay, unit);
+  }
+
+  private Iterable<TwillController> wrapControllers(Iterable<TwillController> controllers, String applicationName) {
+    if (isMasterService(applicationName)) {
+      return controllers;
+    }
+
+    try {
+      final ProgramId programId = TwillAppNames.fromTwillAppName(applicationName);
+      return Iterables.transform(controllers, new Function<TwillController, TwillController>() {
+        @Override
+        public TwillController apply(TwillController controller) {
+          return new ImpersonatedTwillController(controller, impersonator, programId);
+        }
+      });
+    } catch (IllegalArgumentException e) {
+      // If the conversion from twill app name to programId failed, don't wrap
+      return controllers;
+    }
+  }
+
+  private SecureStoreUpdater wrapSecureStoreUpdater(final SecureStoreUpdater updater) {
+    return new SecureStoreUpdater() {
+      @Override
+      public SecureStore update(final String application, final RunId runId) {
+        if (isMasterService(application)) {
+          return updater.update(application, runId);
+        }
+
+        ProgramId programId;
+        try {
+          programId = TwillAppNames.fromTwillAppName(application);
+        } catch (IllegalArgumentException e) {
+          // If the conversion from twill app name to programId failed, just delegate
```

Comment:
It can happen if someone is using Twill directly on the same cluster without going through CDAP. For app that launches through CDAP, this never happen.